### PR TITLE
feat: Add some chain calls for Log facade

### PR DESCRIPTION
**Pull request by kkumar-gcc**
Closes https://github.com/goravel/goravel/issues/240

## 📑 Description
This pull request introduces a set of chainable function calls to enhance the capabilities of the Log facade. These new functions enable more efficient and expressive logging configuration and customization, making it easier to attach relevant information to log entries. The added methods include:
- `User`: Associates a user with the log entry.
- `Owner`: Specifies the responsible colleague or team for handling the logged error, aiding in alerting workflows.
- `Hint`: Provides a hint that can expedite debugging processes.
- `Code`: Set a code or slug that describes the error
- `With`:  Appends key-value pairs to the log entry's contextual information
- `Tags`: Adds tags to categorize and group log entries
- `Request`: Links an HTTP request to the log entry.
- `Response`: Associates an HTTP response with the log entry.
- `In`: Sets the feature category or domain in which the log entry is relevant

```go
//example
facades.Log().
        User("kkumar-gcc").
        Owner("team@goravel.dev").
	Hint("Check the database connection.").
	Code("500").
	With(map[string]interface{}{"key1": "value1", "key2": "value2"}).
        Tags([]string{"error", "critical"}).
	In("backend").
        Errorf("Goravel: %s", "World")

//output(in log file)
[2023-08-28 15:56:15] test.error: Goravel: World {"error":{"message":"Goravel: World","tags":["error","critical"],"hint":"Check the database connection.","user":"john.doe","code":"500","domain":"backend","context":{"key1":"value1","key2":"value2"},"owner":"team@example.com","stacktrace":"Goravel: World\n\tsuite.(*Suite).Run.func2:/home/krish/go/pkg/mod/github.com/stretchr/testify@v1.8.4/suite/suite.go:112\n\tlog.(*LogrusTestSuite).TestLogrus.func26:/home/krish/goravel/framework/log/logrus_writer_test.go:246\n\tlog.(*LogrusTestSuite).TestLogrus.func20:/home/krish/goravel/framework/log/logrus_writer_test.go:200\n\tlog.(*Writer).Errorf:/home/krish/goravel/framework/log/logrus_writer.go:107\n\tlog.(*Writer).withStackTrace:/home/krish/goravel/framework/log/logrus_writer.go:198"}}

```

image
image

## ✅ Checks

- [x] My pull request adheres to the code style of this project
- [x] My code requires changes to the documentation
- [ ] I have updated the documentation as required
- [x] All the tests have passed

### CODE REVIEW

File: contracts/log/log.go [hunk 1]
```diff
@@ -41,6 +43,16 @@ type Writer interface {
 	Fatalf(format string, args ...any)
 	Panic(args ...any)
 	Panicf(format string, args ...any)
+	User(userID string, userData ...map[string]any) Writer
+	Owner(ownerID string) Writer
```
Comment (hwbrzzl) [inline contracts/log/log.go:47]: Ditto

```suggestion
	Owner(owner any) Writer
```

File: contracts/log/log.go [hunk 2]
```diff
@@ -41,6 +43,16 @@ type Writer interface {
...
+	Hint(hint string) Writer
+	Trace(trace string) Writer
```
Comment (hwbrzzl) [inline contracts/log/log.go:49]: Maybe we can remove this method, adding the trace information when `Error` automatically.

File: contracts/log/log.go [hunk 3]
```diff
@@ -41,6 +43,16 @@ type Writer interface {
...
+	Code(code string) Writer
+	With(data map[string]any) Writer
+	Tags(tags []string) Writer
+	Request(req http.Request) Writer
+	Response(res http.Response) Writer
+	In(domain string) Writer
```
Comment (hwbrzzl) [inline contracts/log/log.go:55]: Add description for these new methods like oops.

File: log/logrus_writer.go [hunk 4]
```diff
@@ -2,69 +2,281 @@ package log
...
 
 func NewWriter(instance *logrus.Entry) log.Writer {
-	return &Writer{instance: instance}
+	return &Writer{
+		instance: instance,
+
+		message:  "",
+		code:     "",
+		time:     time.Now(),
+		duration: 0,
```
Comment (hwbrzzl) [inline log/logrus_writer.go:57]: It seems unnecessary.

**Review by hwbrzzl — COMMENTED** [review]
Good job, could you please add some screenshots for this feature?

File: log/formatter/general.go [hunk 5]
```diff
@@ -35,16 +36,107 @@ func (general *General) Format(entry *logrus.Entry) ([]byte, error) {
...
+		Message string   `json:"message"`
+		Stack   []string `json:"stack"`
+	} `json:"root"`
+	Wrap []struct {
+		Message string `json:"message"`
+		Stack   string `json:"stack"`
+	} `json:"wrap"`
+}
+
+func formatStackTraces(stackTraces interface{}) (string, error) {
```
Comment (hwbrzzl) [inline log/formatter/general.go:122]:
```suggestion
func formatStackTraces(stackTraces any) (string, error) {
```

**Review by hwbrzzl — COMMENTED** [review]
Perfect! please add some test cases.

File: log/logrus_writer.go [hunk 6]
```diff
@@ -2,69 +2,267 @@ package log
...
+
+	// user
+	user any
+
+	// http
+	request  http.Request
+	response http.Response
+
+	// stacktrace
+	stackEnabled bool
```
[the remark this answers is not in the document]
Comment (hwbrzzl) [reply log/logrus_writer.go]: :fu:

File: log/logrus_writer_test.go [hunk 7]
```diff
@@ -207,6 +208,123 @@ func (s *LogrusTestSuite) TestLogrus() {
...
+				assert.True(s.T(), file.Contain(dailyLog, "test.info: Goravel\nowner: \"team@goravel.dev\""))
+			},
+		},
+		{
+			name: "Request",
+			setup: func() {
+				mockDriverConfig(mockConfig)
+
+				log = NewLogrusApplication(mockConfig)
+				log.Request(nil).Info("Goravel")
```
Comment (hwbrzzl) [inline log/logrus_writer_test.go:269]: S: Create a TestRequest to test this feature, like: https://github.com/goravel/framework/blob/c38f2bf119f3858ab9b64b20a4199e6f67ca3161/http/middleware/throttle_test.go#L282

File: log/logrus_writer_test.go [hunk 8]
```diff
@@ -207,6 +208,123 @@ func (s *LogrusTestSuite) TestLogrus() {
...
+				assert.True(s.T(), file.Contain(dailyLog, "test.info: Goravel"))
+			},
+		},
+		{
+			name: "Response",
+			setup: func() {
+				mockDriverConfig(mockConfig)
+
+				log = NewLogrusApplication(mockConfig)
+				log.Response(nil).Info("Goravel")
```
Comment (hwbrzzl) [inline log/logrus_writer_test.go:282]: Ditto

File: log/formatter/general.go
Comment (hwbrzzl) [inline log/formatter/general.go:1]: Could you please add some test cases for this file?

File: log/formatter/general_test.go
Comment (hwbrzzl) [inline log/formatter/general_test.go:1]: S: We should add test cases for different functions, like: `TestGeneral_Format`, `TestFormatStackTraces`, `TestDeleteKey`, `TestFormatData`.

File: log/formatter/general_test.go [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+	"strings"
+	"testing"
+
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/suite"
+
+	configmock "github.com/goravel/framework/contracts/config/mocks"
+)
+
+type GeneralFormatterTestSuite struct {
```
Comment (hwbrzzl) [inline log/formatter/general_test.go:13]:
```suggestion
type GeneralTestSuite struct {
```

File: log/formatter/general_test.go [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+	suite.Suite
+	mockConfig  *configmock.Config
+	entry       *logrus.Entry
+	stackTraces any
+}
+
+func TestGeneralFormatterTestSuite(t *testing.T) {
```
Comment (hwbrzzl) [inline log/formatter/general_test.go:20]:
```suggestion
func TestGeneralTestSuite(t *testing.T) {
```

File: log/formatter/general_test.go [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+	s.entry = &logrus.Entry{
+		Level:   logrus.InfoLevel,
+		Message: "Test Message",
+	}
+
+	s.mockConfig.On("GetString", "app.timezone").Return("UTC")
+	s.mockConfig.On("GetString", "app.env").Return("test")
+}
+
+func (s *GeneralFormatterTestSuite) TestGeneral_Format() {
```
Comment (hwbrzzl) [inline log/formatter/general_test.go:35]:
```suggestion
func (s *GeneralFormatterTestSuite) TestFormat() {
```

File: log/formatter/general_test.go [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+	general := NewGeneral(s.mockConfig)
+	formatLog, err := general.Format(s.entry)
+	s.Nil(err)
+	s.NotNil(formatLog)
+}
+
+func (s *GeneralFormatterTestSuite) TestFormatData() {
```
Comment (hwbrzzl) [inline log/formatter/general_test.go:42]: The `formatData` method is a independent method, you don't need to set the test case to `GeneralFormatterTestSuite`.

 ```suggestion
func TestFormatData() {
```

File: log/formatter/general_test.go [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+
+	for _, test := range tests {
+		s.Run(test.name, func() {
+			test.setup()
+			test.assert()
+		})
+	}
+}
+
+func (s *GeneralFormatterTestSuite) TestDeleteKey() {
```
Comment (hwbrzzl) [inline log/formatter/general_test.go:105]: Ditto

```suggestion
func TestDeleteKey() {
```

File: log/formatter/general_test.go [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+	}
+
+	for _, test := range tests {
+		s.Run(test.name, func() {
+			test.assert()
+		})
+	}
+}
+
+func (s *GeneralFormatterTestSuite) TestFormatStackTraces() {
```
Comment (hwbrzzl) [inline log/formatter/general_test.go:149]: Ditto

```suggestion
func TestFormatStackTraces() {
```

File: log/formatter/general_test.go [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+
+func (s *GeneralFormatterTestSuite) SetupTest() {
+	s.mockConfig = &configmock.Config{}
+	s.entry = &logrus.Entry{
+		Level:   logrus.InfoLevel,
+		Message: "Test Message",
+	}
+
+	s.mockConfig.On("GetString", "app.timezone").Return("UTC")
+	s.mockConfig.On("GetString", "app.env").Return("test")
```
Comment (hwbrzzl) [inline log/formatter/general_test.go:32]: S: Move these lines to `TestFormat`, because they belong to it.

File: log/formatter/general_test.go [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+}
+
+func (s *GeneralFormatterTestSuite) TestGeneral_Format() {
+	general := NewGeneral(s.mockConfig)
+	formatLog, err := general.Format(s.entry)
+	s.Nil(err)
+	s.NotNil(formatLog)
```
Comment (hwbrzzl) [inline log/formatter/general_test.go:39]:
```suggestion
	s.NotNil(formatLog)
	
	s.mockConfig.AssertExpectations(t)
```

File: log/formatter/general.go [hunk 17]
```diff
@@ -35,16 +36,107 @@ func (general *General) Format(entry *logrus.Entry) ([]byte, error) {
...
-	var newLog string
+	b.WriteString(fmt.Sprintf("[%s] %s.%s: %s\n", timestamp, general.config.GetString("app.env"), entry.Level, entry.Message))
+	data := entry.Data
+	if len(data) > 0 {
+		formattedData, err := formatData(data)
+		if err != nil {
+			return nil, err
+		}
+		b.WriteString(formattedData)
+	}
```
Comment (hwbrzzl) [inline log/formatter/general.go:47]: S: Add two test cases(normal and error) for these lines.

File: log/formatter/general_test.go (hunk 16 above)
Comment (hwbrzzl) [inline log/formatter/general_test.go:39]: Q: Can we judge a specific content?

File: log/formatter/general_test.go [hunk 18]
```diff
@@ -0,0 +1,209 @@
...
+	}{
+		{
+			name: "Data is empty",
+			setup: func() {
+				s.entry.Data = logrus.Fields{}
+			},
+			assert: func() {
+				formattedData, err := formatData(s.entry.Data)
+				s.Nil(err)
+				s.Empty(formattedData)
```
Comment (hwbrzzl) [inline log/formatter/general_test.go:56]: Q: Can we judge a specific content?

File: log/logrus_writer_test.go [hunk 19]
```diff
@@ -283,3 +430,259 @@ func mockDriverConfig(mockConfig *configmock.Config) {
...
+func (r *TestRequest) Method() string {
+	return "GET"
+}
+
+func (r *TestRequest) Path() string {
+	return "/test"
+}
+
+func (r *TestRequest) Url() string {
+
```
Comment (hwbrzzl) [inline log/logrus_writer_test.go:457]: S: Remove these empty lines.

File: log/logrus_writer_test.go [hunk 20]
```diff
@@ -283,3 +430,259 @@ func mockDriverConfig(mockConfig *configmock.Config) {
...
+type TestRequest struct{}
+
+func (r *TestRequest) Header(key string, defaultValue ...string) string {
+
+	panic("do not need to implement it")
+}
+
+func (r *TestRequest) Headers() nethttp.Header {
+	return nethttp.Header{
+		"Sec-Fetch-User": []string{"?1"},
```
Comment (hwbrzzl) [inline log/logrus_writer_test.go:443]: Q: Why have a `?`?

File: log/logrus_writer_test.go [hunk 21]
```diff
@@ -283,3 +430,259 @@ func mockDriverConfig(mockConfig *configmock.Config) {
 	mockConfig.On("GetString", "app.timezone").Return("UTC")
 	mockConfig.On("GetString", "app.env").Return("test")
 }
+
+type TestRequest struct{}
+
+func (r *TestRequest) Header(key string, defaultValue ...string) string {
+
```
Comment (hwbrzzl) [inline log/logrus_writer_test.go:437]: S: Remove these empty lines.

**Review by hwbrzzl — COMMENTED** [review]
Pretty good

File: log/logrus_writer_test.go (hunk 20 above)
Comment (kkumar-gcc) [reply log/logrus_writer_test.go:443]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Sec-Fetch-User
image

File: log/formatter/general_test.go (hunk 18 above)
Comment (kkumar-gcc) [reply log/formatter/general_test.go:56]: `formatData` will return formatted String so if we want to judge a specific content, ig we can use check that in formattedData

File: log/formatter/general_test.go [hunk 22]
```diff
@@ -0,0 +1,294 @@
...
+		})
+	}
+
+	s.mockConfig.AssertExpectations(s.T())
+}
+
+func TestFormatData(t *testing.T) {
+	if testing.Short() {
+		t.Skip("Skipping tests of using docker")
+	}
```
Comment (hwbrzzl) [inline log/formatter/general_test.go:95]: S: This judgment is only used in the Windows system, this is unnecessary if this case can be tested in Windows. The same as below.

**Review by hwbrzzl — COMMENTED** [review]
You have a big improvement, Bro.

File: log/formatter/general_test.go [hunk 23]
```diff
@@ -0,0 +1,281 @@
...
+	suite.Suite
+	mockConfig *configmock.Config
+	entry      *logrus.Entry
+}
+
+func TestGeneralTestSuite(t *testing.T) {
+	if testing.Short() {
+		t.Skip("Skipping tests of using docker")
+	}
+
```
Comment (hwbrzzl) [inline log/formatter/general_test.go:24]: S: Remove these lines.

**Review by hwbrzzl — APPROVED** [review]
Great job!

**Pull request merged**